[210] Add ability for leaders to select suites

esoterik opened the pull request:
Resolves #210
  - Add GroupSuiteSelector

orenyk commented on app/controllers/groups_controller.rb:
```diff
@@ -103,6 +105,16 @@ def assign_lottery
   end
 
+  def select_suite
+    @suites = @draw.suites.where(size: @group.size)
```
`@draw.suites.available`

orenyk commented on app/policies/group_policy.rb:
```diff
@@ -83,6 +83,15 @@ def change_leader?
 
+  def select_suite?
+    user.admin? || (user.leader_of?(record) &&
+                    record.draw.next_groups.any? { |g| g == record })
```
Why don't we move this to a model method on `draw`, e.g. `Draw#next_group?`?

orenyk commented on app/services/group_suite_selector.rb:
```diff
@@ -0,0 +1,53 @@
+    @mailer = mailer
+  end
+
+  # Select / assign a suite to a group. Checks to make sure that the grou does
```
typo, `grou`

orenyk commented on spec/features/groups/suite_selection_spec.rb:
```diff
@@ -0,0 +1,22 @@
+RSpec.feature 'Suite Selection' do
+  let(:leader) do
+    FactoryGirl.create(:draw_in_selection).next_groups.first.leader.tap do |l|
+      l.update(password: 'password')
```
?

esoterik replied:
no idea but we can't log in with the user otherwise.

orenyk commented on spec/services/group_suite_selector_spec.rb:
```diff
@@ -0,0 +1,60 @@
+        group = instance_spy('group', suite: nil, draw: draw)
+        suite = mock_suite(id: 123, present: false)
+        result = described_class.select(group: group, suite_id: suite.id.to_s)
+        expect(result[:msg].keys).to include(:error)
```
I thought you wanted to check exclusivity here, e.g. `match([:error])`?

esoterik replied:
oh yeah you're right

esoterik replied:
realized that I used `include` because I copied this from `SuiteSelector` spec -- changing it there as well

esoterik replied:
also we want `eq`, not `match`

orenyk commented on spec/services/group_suite_selector_spec.rb:
```diff
@@ -0,0 +1,60 @@
+        group = instance_spy('group', suite: nil, draw: draw)
+        suite = mock_suite(id: 123)
+        described_class.select(group: group, suite_id: suite.id.to_s)
+        expect(group.draw).to have_received(:notify_next_groups)
```
Do we want to check that it passes `next_group.leader`?

orenyk commented on spec/services/group_suite_selector_spec.rb:
```diff
@@ -0,0 +1,60 @@
+    context' success' do
+      # rubocop:disable RSpec/ExampleLength
+      it 'sends emails to the next groups' do
+        next_group = instance_spy('Group', lottery_number: 12)
```
Why does this need a lottery number?

esoterik replied:
it only sends emails when the lottery number is different than the lottery number of the current group

orenyk replied:
Can we be consistent about the capitalization of class names as passed to `instance_spy`? I think we should stick with camel case since that fixed that weird test before, but either way we should be consistent within a given test.

orenyk commented on spec/services/group_suite_selector_spec.rb:
```diff
@@ -0,0 +1,60 @@
+      it 'sends emails to the next groups' do
+        next_group = instance_spy('Group', lottery_number: 12)
+        draw = instance_spy('Draw', next_groups: [next_group])
+        group = instance_spy('group', suite: nil, draw: draw)
```
both `group` and `suite` could be set in `let` blocks to make the tests more concise.

orenyk commented on spec/services/group_suite_selector_spec.rb:
```diff
@@ -0,0 +1,60 @@
+        group = instance_spy('group', suite: nil, draw: draw)
+        suite = mock_suite(id: 123)
+        result = described_class.select(group: group, suite_id: suite.id.to_s)
+        expect(result[:msg].keys).to include(:success)
```
same here, `match` vs `include`.

orenyk commented on app/policies/group_policy.rb:
```diff
@@ -83,6 +83,15 @@ def change_leader?
     edit?
   end
 
+  def select_suite?
```
So this policy will change in #276? Or should we restrict selection to groups that are the next groups in their respective draws?

esoterik replied:
I think it would change -- reps can assign as long as the group is in the next groups

orenyk commented on spec/models/draw_spec.rb:
```diff
@@ -255,10 +255,23 @@
+      draw = FactoryGirl.build_stubbed(:draw)
+      group = instance_spy('Group', leader: instance_spy('User'))
+      allow(draw).to receive(:next_groups).and_return([group])
+      mailer = instance_spy('StudentMailer')
```
ummm... changing this to `instance_spy('student_mailer')` passes this test... not sure what's going on. I'll leave it to you to update and rebase.

esoterik commented:
@orenyk ready for rereview! will squash and rebase after this one

esoterik commented:
no, i reverted them because that's the file that caused the segfault when i changed it; i'll see what i can do to avoid the segfault and fix them as much as possible -- but i'm not super concerned because it's essentially just cleaning up tests

orenyk commented:
Oh ok, I wasn't sure. It's fine to just skip it all if you want then if those changes are causing problems - there will still be the merge conflicts to address.

orenyk commented:
Failing test? Also, might as well rebase and solve merge conflicts before a final review.

orenyk commented on app/models/draw.rb:
```diff
@@ -152,7 +152,24 @@ def size_locked?(size)
+  #
+  # @param [Mailer] The mailer to use
+  def notify_next_groups(mailer = StudentMailer)
+    college = College.first
```
This can just be `@current_college` I think.

orenyk commented on app/models/user.rb:
```diff
@@ -88,6 +88,13 @@ def pretty_intent
+  # Returns true if the user is the leader of the given group
+  #
+  # @return [Boolean]
+  def leader_of?(group)
```
Ooo I feel like we could heavily refactor `GroupPolicy` and `DrawlessGroupPolicy`, if not `DrawPolicy. I'll open an issue for post-pilot.

orenyk commented on app/services/group_suite_selector.rb:
```diff
@@ -0,0 +1,53 @@
+  # @param suite_id [Integer] the suite id to assign to the group
+  # @param mailer [Mailer] the mailer to use to email next groups
+  def initialize(group:, suite_id:, mailer: StudentMailer)
+    @group = group
```
Can we use `super` here instead for the first three instance variables?

orenyk commented on app/views/groups/select_suite.html.erb:
```diff
@@ -0,0 +1,5 @@
+<h1>Select a Suite</h1>
+<%= simple_form_for [@draw, @group], url: assign_suite_draw_group_path(@draw, @group) do |f| %>
+  <%= f.input :suite, collection: @suites, as: :radio_buttons, label_method: :number %>
+  <%= f.submit "Submit Selection" %>
```
Do we want to add a tooltip here? Maybe something about how this is final and commits them to living on campus?
